domain/machine/state/state.go

```diff
@@ -857,3 +858,89 @@ WHERE  name = $machineName.name`
 		return nil
 	})
 }
+
+// LXDProfiles returns the names of the LXD profiles on the machine.
+func (st *State) LXDProfiles(ctx context.Context, mUUID string) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineUUIDQuery := machineUUID{UUID: mUUID}
+	query := `
+SELECT &lxdProfile.name
+FROM   lxd_profile
+WHERE  machine_uuid = $machineUUID.uuid`
+	queryStmt, err := st.Prepare(query, lxdProfile{}, machineUUIDQuery)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var result []lxdProfile
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, queryStmt, machineUUIDQuery).GetAll(&result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return internalerrors.Errorf("retrieving lxd profiles for machine %q: %w", mUUID, err)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	lxdProfiles := make([]string, len(result))
+	for i, lxdProfile := range result {
+		lxdProfiles[i] = lxdProfile.Name
+	}
+	return lxdProfiles, nil
+}
+
+// SetLXDProfiles adds the list of LXD profile names to the lxd_profile table
+// for the given machine.
+// If one of the profile names already exists for the given machine, then it
+// will be kept without returning an error.
+func (st *State) SetLXDProfiles(ctx context.Context, mUUID string, profileNames []string) error {
+	if len(profileNames) == 0 {
+		return nil
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return internalerrors.Errorf("cannot get database to set lxd profiles %v for machine %q: %w", profileNames, mUUID, err)
+	}
+
+	checkMachineExistsStmt, err := st.Prepare(`
+SELECT true AS &machineExists.exists
+FROM   machine
+WHERE  machine.uuid = $machineUUID.uuid`, machineUUID{}, machineExists{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setLXDProfileStmt, err := st.Prepare(`
+INSERT INTO lxd_profile (*)
+VALUES      ($lxdProfile.*)
+ON CONFLICT DO NOTHING`, lxdProfile{})
```

Comment:
I think the better approach is to ask which profiles haven't been inserted, and then insert the ones that are missing. This would work in our favor if there are profiles that have already been applied. You'd do nothing. Also, the number of inserts will be less.

Always prefer a read over a write. The fewer writes we have, the less contention we have, which means we're more likely to go brrrrrrrrrrrrr!